04-onion-routing.md

```diff
@@ -248,18 +248,104 @@ This is a more flexible format, which avoids the redundant `short_channel_id` fi
     1. type: 6 (`short_channel_id`)
     2. data:
         * [`short_channel_id`:`short_channel_id`]
+    1. type: 8 (`total_msat`)
+    2. data:
+        * [`tu64`:`total_msat`]
```

Comment:
Today t-bast and I discussed the tlv payload a bit more with the idea in mind to consider the receiver id as a more generic `payment_address` (taken from #658). This is a property that has meaning for any type of payment. It is like a box on the destination node in which the payment is dropped. So the destination of a Lightning payment would consist of a node pubkey and a payment address. 

Depending on the type of payment, there may be additional fields required. For an mpp payment there is a specific `payment_hash` that needs to be used on the htlc to get a proof of payment for the address that is being paid to. For an amp payment, there is a `set_id` and `shares`.

For mpp, the `payment_address` is single use and needs to be generated randomly. But for amp or a different form of spontaneous payment, it could actually be something that is derived from other data. For example, the hash of a user name. Users could have their own deposit boxes on a node and people donating don't need an invoice to donate to a specific user. They just include the user name hash as the payment address in the payload.

As we don't know yet exactly how this `payment_address` will be used, it is probably safer to use our standard 32 byte length for it. If it will be used with a hash, it will provide sufficient collision resistance. And there may also be other forms of composition that benefit from having this standard size.

Reply:
tl;dr would be to change the tlv to:

1. type: 8 (`basic_mpp`)
2. data:
    * [`32*byte`:`payment_addr`]
    * [`u64`:`total_msat`]

And add this `payment_addr` to the invoice (needs some changes in Bolt 11 similar to the ones I made in #671)

Reply:
I assumed we'd add some nonce on top of this, yes.

But for my "offers" proposal (still, mainly in my head!), the nonce would be a merkle tree, so 32-bytes is a nice number.  In simplest incarnation (base AMP), it's provided in the invoice.

I'll add it.

Reply:
Good point, 32 bytes is interesting because it fits a merkle root which opens the door for many applications.

Reply:
I made it variable, because why not?

Rebased your PR at the end with minor changes to make it fit.